File: cucumber-expressions/javascript/src/combinatorial_generated_expression_factory.js
const GeneratedExpression = require('./generated_expression')

class CombinatorialGeneratedExpressionFactory {
  constructor(expressionTemplate, parameterTypeCombinations) {
    this._expressionTemplate = expressionTemplate
    this._parameterTypeCombinations = parameterTypeCombinations
  }

  generateExpressions() {
    const generatedExpressions = []
    this._generatePermutations(generatedExpressions, 0, [])
    return generatedExpressions
  }

  _generatePermutations(generatedExpressions, depth, currentParameterTypes) {
    if (depth === this._parameterTypeCombinations.length) {
      generatedExpressions.push(new GeneratedExpression(this._expressionTemplate, currentParameterTypes))
      return
    }

    for (let i = 0; i < this._parameterTypeCombinations[depth].length; ++i) {
      const newCurrentParameterTypes = currentParameterTypes.slice() // clone
      newCurrentParameterTypes.push(this._parameterTypeCombinations[depth][i])
      this._generatePermutations(
        generatedExpressions,
        depth + 1,
        newCurrentParameterTypes
      )
    }
  }
}

module.exports = CombinatorialGeneratedExpressionFactory

File: cucumber-expressions/go/combinatorial_generated_expression_factory.go
package cucumberexpressions

// 256 generated expressions ought to be enough for anybody
const maxExpressions = 256

type CombinatorialGeneratedExpressionFactory struct {
	expressionTemplate        string
	parameterTypeCombinations [][]*ParameterType
}

func NewCombinatorialGeneratedExpressionFactory(expressionTemplate string, parameterTypeCombinations [][]*ParameterType) *CombinatorialGeneratedExpressionFactory {
	return &CombinatorialGeneratedExpressionFactory{
		expressionTemplate:        expressionTemplate,
		parameterTypeCombinations: parameterTypeCombinations,
	}
}

func (c *CombinatorialGeneratedExpressionFactory) GenerateExpressions() []*GeneratedExpression {
	generatedExpressions := &GeneratedExpressionList{}
	c.generatePermutations(generatedExpressions, 0, nil)
	return generatedExpressions.ToArray()
}

func (c *CombinatorialGeneratedExpressionFactory) generatePermutations(generatedExpressions *GeneratedExpressionList, depth int, currentParameterTypes []*ParameterType) {
	if len(generatedExpressions.elements) >= maxExpressions {
		return
	}

	if depth == len(c.parameterTypeCombinations) {
		generatedExpressions.Push(
			NewGeneratedExpression(c.expressionTemplate, currentParameterTypes),
		)
		return
	}

	for _, parameterType := range c.parameterTypeCombinations[depth] {
		// Avoid recursion if no elements can be added.
		if len(generatedExpressions.elements) >= maxExpressions {
			return
		}

		c.generatePermutations(
			generatedExpressions,
			depth+1,
			append(currentParameterTypes, parameterType),
		)
	}
}

type GeneratedExpressionList struct {
	elements []*GeneratedExpression
}

func (g *GeneratedExpressionList) Push(expr *GeneratedExpression) {
	g.elements = append(g.elements, expr)
}

func (g *GeneratedExpressionList) ToArray() []*GeneratedExpression {
	return g.elements
}
